v3/generate.go

```diff
@@ -0,0 +1,233 @@
+			}
+			operation["parameters"] = paramsForClient
+		}
+	}
+
+	definitions := m["definitions"].(map[interface{}]interface{})
+	// Remove any definitions that are used only for Events
+	for nameInterface, definition := range definitions {
+
+		name := nameInterface.(string)
+		if strings.HasSuffix(name, ".created") ||
+			strings.HasSuffix(name, ".updated") ||
+			strings.HasSuffix(name, ".deleted") ||
+			strings.HasSuffix(name, "Object") {
+			delete(definitions, nameInterface)
+			continue
+		}
+
+		if strings.HasPrefix(name, "Event") {
+			delete(definitions, nameInterface)
+			continue
+		}
+
+		modifyDefinitions(version, false, name, definition.(map[interface{}]interface{}))
+	}
+
+	return yaml.Marshal(m)
+}
+
+// generateEventsAPIYml generates the events API from the base yml for a specific version. It does
+// this by removing things from the yml, for example the non /events endpoints.
+func generateEventsAPIYml(i map[interface{}]interface{}, version string) ([]byte, error) {
+	m := deepCopyMap(i)
+
+	m["basePath"] = "/" + version
+	info := m["info"].(map[interface{}]interface{})
+	info["title"] = "Events API"
+	info["description"] = "The Clever Events API"
+	info["version"] = strings.Replace(version, "v", "", -1) + ".0"
+
+	paths := m["paths"].(map[interface{}]interface{})
+	for path := range paths {
+		if !strings.Contains(path.(string), "/events") {
+			delete(paths, path)
+			continue
+		}
+	}
+
+	definitions := m["definitions"].(map[interface{}]interface{})
+	for nameInterface, definition := range definitions {
+		name := nameInterface.(string)
+		modifyDefinitions(version, false, name, definition.(map[interface{}]interface{}))
+	}
+
+	return yaml.Marshal(m)
+}
+
+// generateClientYml generates the yml for the client libraries. It removes things we don't need
+// implementations to use.
+func generateClientYml(i map[interface{}]interface{}, versionStr string) ([]byte, error) {
+	m := deepCopyMap(i)
+
+	delete(m, "x-sample-languages")
+	info := m["info"].(map[interface{}]interface{})
+	info["title"] = "Clever API"
+	info["description"] = "The Clever API"
+	m["basePath"] = "/" + versionStr
+
+	paths := m["paths"].(map[interface{}]interface{})
+	for path, methodOp := range paths {
+		for _, o := range methodOp.(map[interface{}]interface{}) {
+			operation := o.(map[interface{}]interface{})
+
+			// Tweak the tags so they show up correctly in the client libraries
+			if strings.Contains(path.(string), "/events") {
+				operation["tags"] = []string{"Events"}
+			} else {
+				operation["tags"] = []string{"Data"}
+			}
+		}
+	}
+
+	definitions := m["definitions"].(map[interface{}]interface{})
+	for name, definition := range definitions {
+		modifyDefinitions(versionStr, true, name.(string), definition.(map[interface{}]interface{}))
+	}
+
+	return yaml.Marshal(m)
+}
+
+// deepCopyMap recursively copies the map. We use this so we can modify it for each output yml
+func deepCopyMap(m map[interface{}]interface{}) map[interface{}]interface{} {
```

Comment:
we should probably have a helper library for some of these since they're shared across versions

Reply:
Created 🥳